### PR TITLE
GPU device port to use resources package

### PR DESCRIPTION
**Pull request by tomponline**
Updates GPU device to use resources package for probing info about the GPUs available on the host.

Includes  https://github.com/lxc/lxd/pull/6100

### CODE REVIEW

File: lxd/device/gpu.go [hunk 1]
```diff
@@ -151,66 +79,91 @@ func (d *gpu) Start() (*RunConfig, error) {
...
-		}
+		// Handle DRM devices if present and matches criteria.
+		if gpu.DRM != nil && (d.config["id"] == "" || fmt.Sprintf("%d", gpu.DRM.ID) == d.config["id"]) {
+			found = true
 
-		if !gpu.isNvidia {
-			continue
-		}
+			// DRM &{ID:0 CardName:card0 CardDevice:226:0 ControlName: ControlDevice: RenderName:renderD128 RenderDevice:226:128}
+			// TODO: Need to also support ControlName & ControlDevice?
```
Comment (tomponline) [inline lxd/device/gpu.go]: @stgraber I have a question regarding the DRM struct that is returned from the resources package. It includes `ControlName` and `ControlDevice` fields that were empty on my test system (even with an nvidia card). Should these be checked for being populated and used to setup a unix char device like the CardName and RenderName fields?

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/device/gpu.go (hunk 1 above)
Comment (stgraber) [reply lxd/device/gpu.go]: Ah, I do have it on Intel here:

```
  DRM:
    ID: 0
    Card: card0 (226:0)
    Control: controlD64 (226:0)
    Render: renderD128 (226:128)
```

But indeed the file itself in /dev/dri doesn't exist despite the /sys entry existing.
I'd say that we should get the 3 files from the DRM struct, if they're not empty in the struct and exist in /dev/dri on the host, then pass them to the container.

I suspect it's a kernel version thing, some older kernels may have had the control device whereas newer ones just skip over that particular one in favor of card?

Comment (tomponline) [reply lxd/device/gpu.go]: @stgraber ok thanks will do.

I have another question, the old implementation allowed for detecting nvidia cards and passing through the /dev/nvidiaX and other /dev/nvidia-* devices into the container without `libnvidia-container-tools` being installed. Whereas now if you don't have `libnvidia-container-tools` installed you will only get the `/dev/dri` devices. Is this intentional?

Comment (stgraber) [reply lxd/device/gpu.go]: I think that should be fine as the LXD snap ships with the needed bits for that.

If it becomes a problem, I'd rather we add logic to the resources API to fill the NVIDIA struct when we don't have the nvidia tools nearby.

Comment (tomponline) [reply lxd/device/gpu.go]: @stgraber I've changed my BIOS settings so both the nvidia GPU and the onboard intel GPU are active now, so I was able to test the same scenario you saw of getting `Control` property without an actual file in /dev/dri existing.

Comment (tomponline) [reply lxd/device/gpu.go]: @stgraber ok thanks, I'll update the tests so they only do nvidia tests of the tool is installed.

**Comment by tomponline** [comment]
@stgraber ready for review

**Comment by stgraber** [comment]
Rebased following the test branch getting merged. Will review once jenkins is happy.

**Pull request merged**